pallets/token-factory/src/lib.rs

```diff
@@ -0,0 +1,329 @@
+// Copyright 2019-2020 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Pallet for registering, minting, and burning ERC20 tokens (in EVM from runtime)
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::pallet;
+
+pub use pallet::*;
+
+#[pallet]
+pub mod pallet {
+	use ethereum_types::BigEndianHash;
+	use frame_support::{pallet_prelude::*, traits::Currency};
+	use frame_system::pallet_prelude::*;
+	use parity_scale_codec::FullCodec;
+	use rustc_hex::FromHex;
+	use sp_core::{H160, H256, U256};
+	use sp_runtime::{
+		traits::{AtLeast32BitUnsigned, Convert, Zero},
+		SaturatedConversion,
+	};
+	use sp_std::fmt::Debug;
+
+	/// ERC20PresetMinterBurner contract bytecode
+	const CONTRACT_BYTECODE: &str = include_str!("../contract/bytecode.txt");
+	type BalanceOf<T> =
+		<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+	/// The ERC token factory pallet
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Configuration trait of this pallet.
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_ethereum::Config + pallet_sudo::Config {
+		/// Overarching event type
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// Balances type
+		type Currency: Currency<Self::AccountId>;
+		/// ERC token identifier
+		type TokenId: Clone + Copy + AtLeast32BitUnsigned + FullCodec + Debug;
+		/// Convert from AccountId to H160, should be identity map for Moonbeam
+		type AccountToH160: Convert<Self::AccountId, H160>;
```

Comment:
If you couple to pallet evm, you can use the same mapping that is in its config trait without needing this associated type.

Reply:
I found this associated type in pallet_evm

```rust
/// Mapping from address to account id.
type AddressMapping: AddressMapping<Self::AccountId>;
```

but I did not find account id to address, which is what I need. It's the same identity map, but the direction matters because I need to convert from T::AccountId to H160

Reply:
I guess I could just do this for the two times I need this conversion
```rust
T::AccountId::decode(&mut &addr[..]).expect("AccountId: H160")
```